File: dvbViewerController/src/main/java/org/dvbviewer/controller/io/data/FavMatcher.java
package org.dvbviewer.controller.io.data;

import android.support.annotation.Nullable;

import org.dvbviewer.controller.entities.Channel;
import org.dvbviewer.controller.entities.ChannelGroup;
import org.dvbviewer.controller.entities.ChannelRoot;

import java.util.ArrayList;
import java.util.List;

/**
 * Created by rbaun on 08.01.16.
 */
public class FavMatcher {

    public List<ChannelGroup> matchFavs(List<ChannelRoot> channelRoots, @Nullable List<ChannelGroup> favList){
        List<ChannelGroup> result = new ArrayList<>();
        if (favList != null){
            for (ChannelGroup favGroup : favList){
                ChannelGroup tmp = getCurrentFavGroup(result, favGroup);
                for (Channel.Fav fav : favGroup.getFavs()){
                    Channel chan = getMatchedChannel(channelRoots, fav.id);
                    if (chan != null){
                        chan.setFavPosition(fav.position);
                        chan.setFlag(Channel.FLAG_FAV);
                        tmp.getChannels().add(chan);
                    }
                }
            }
        }
        return result;
    }

    @Nullable
    private Channel getMatchedChannel(@Nullable List<ChannelRoot> channelRoots, long favId) {
        if (channelRoots != null){
            for(ChannelRoot roots : channelRoots){
                for( ChannelGroup group : roots.getGroups()){
                    for(Channel chan : group.getChannels()){
                        if (chan.getChannelID() == favId){
                            return chan;
                        }
                    }
                }
            }
        }
        return null;
    }

    private ChannelGroup getCurrentFavGroup(List<ChannelGroup> result, ChannelGroup group) {
        int groupIndex = result.indexOf(group);
        ChannelGroup tmp;
        if (groupIndex >= 0){
            tmp = result.get(groupIndex);
        }else{
            tmp = new ChannelGroup();
            tmp.setName(group.getName());
            tmp.setType(ChannelGroup.TYPE_FAV);
            result.add(tmp);
        }
        return tmp;
    }

}

File: dvbViewerController/src/main/java/org/dvbviewer/controller/data/api/handler/FavMatcher.kt
package org.dvbviewer.controller.data.api.handler

import org.dvbviewer.controller.data.entities.Channel
import org.dvbviewer.controller.data.entities.ChannelGroup
import org.dvbviewer.controller.data.entities.ChannelRoot
import java.util.*

class FavMatcher {

    fun matchFavs(channelRoots: List<ChannelRoot>, favList: List<ChannelGroup>?): List<ChannelGroup> {
        val result = ArrayList<ChannelGroup>()
        if (favList != null) {
            for (favGroup in favList) {
                val tmp = getCurrentFavGroup(result, favGroup)
                for (fav in favGroup.favs) {
                    val chan = getMatchedChannel(channelRoots, fav.id)
                    if (chan != null) {
                        chan.favPosition = fav.position
                        chan.setFlag(Channel.FLAG_FAV)
                        tmp.channels.add(chan)
                    }
                }
            }
        }
        return result
    }

    private fun getMatchedChannel(channelRoots: List<ChannelRoot>?, favId: Long): Channel? {
        if (channelRoots != null) {
            for (roots in channelRoots) {
                for (group in roots.groups) {
                    for (chan in group.channels) {
                        if (chan.channelID == favId) {
                            return chan
                        }
                    }
                }
            }
        }
        return null
    }

    private fun getCurrentFavGroup(result: MutableList<ChannelGroup>, group: ChannelGroup): ChannelGroup {
        val groupIndex = result.indexOf(group)
        val tmp: ChannelGroup
        if (groupIndex >= 0) {
            tmp = result[groupIndex]
        } else {
            tmp = ChannelGroup()
            tmp.name = group.name
            tmp.type = ChannelGroup.TYPE_FAV
            result.add(tmp)
        }
        return tmp
    }

}
